### PR TITLE
Deep linking refactor and DL support in D2L

### PR DESCRIPTION
**Pull request by marcospri**
One step towards, still a work in progress

- https://github.com/hypothesis/lms/issues/5543

Unblocks:

- https://github.com/hypothesis/lms/issues/5544

## Summary

- D2L doesn't allow us to submit a launch URL with query params as part of the return message in deep linking.
- Canvas doesn't fully support (it doesn't send them back) custom parameters in the DL response message but it does allow for them to be part of our DL message.

Based on this: 
- We refactor into a plugin the difference in the launch URL.
- We send the assignment configuration as custom parameters in all LMSes
- We refactor the get_document_url in canvas to explicitly read if from the DL configuration (which happens to be in the query params of the URL).
- In D2L we read the URL form the DL params as a fallback. In this case these params are from the custom parameters.

This PR doesn't:

- Support editing, group sets, etc in D2L deep linking. 
- Create a DeepLinkingPlugin or similar. I rather finish the interface over in Misc and do a move & boilerplate PR then. 

## Testing

- Existing deep linked assignments work as before:

https://hypothesis.instructure.com/courses/125/assignments/873

- Re-deep linking an assignment works too:

https://hypothesis.instructure.com/courses/125/assignments/4682/edit?name=Testing+creation+-+marcos&due_at=null&points_possible=0

- Existing DB backed assignment in D2L also work as before (make sure `make devdata` is up to date):

https://aunltd.brightspacedemo.com/d2l/le/content/6782/viewContent/2424/View

- Test deep linking in D2L, just creating and launch work, not editing, nor group sets:

  - Login in D2L as `HypothesisEng.Teacher`
  - Go to the content section of the test course: 
  https://aunltd.brightspacedemo.com/d2l/le/content/6782/Home?itemIdentifier=TOC
  - Click `Existing activities`  and select `Hypothesis link (deeplinking)` form the menu
  - Pick a D2L file
  - An assignment will be created with the name `https://localhost:48001/lti_launches` until (#5544 is fixed).
  - Launch that assignment.

### CODE REVIEW

File: lms/views/lti/deep_linking.py [hunk 1]
```diff
@@ -179,22 +185,19 @@ def file_picker_to_form_fields_v11(self):
...
     @staticmethod
-    def _get_content_url(request):
-        """
-        Translate content information from the frontend to a launch URL.
-
-        We submit the content information to the LMS as an URL pointing to our
-        `lti_launches` endpoint with any information required to identity
-        the content as query parameters.
-        """
+    def _get_assignment_configuration(request) -> dict:
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:199]: This method changes from both getting the config from the FE and encoding it in an URL to just the the config from the FE.

File: lms/views/lti/deep_linking.py [hunk 2]
```diff
@@ -123,7 +123,13 @@ def file_picker_to_form_fields_v13(self):
             "https://purl.imsglobal.org/spec/lti/claim/message_type": "LtiDeepLinkingResponse",
             "https://purl.imsglobal.org/spec/lti/claim/version": "1.3.0",
             "https://purl.imsglobal.org/spec/lti-dl/claim/content_items": [
-                {"type": "ltiResourceLink", "url": document_url}
+                {
+                    "type": "ltiResourceLink",
+                    "url": self.misc_plugin.get_deeplinking_launch_url(
+                        self.request, assignment_configuration
+                    ),
+                    "custom": assignment_configuration,
```
Comment (marcospri) [inline lms/views/lti/deep_linking.py:131]: Send the configuration for all DL assignments. We'll only use it D2L now but no need to make it D2L specific and bring a plugin method here.

This is part of the spec https://www.imsglobal.org/spec/lti-dl/v2p0#deep-linking-response-example

File: lms/product/plugin/misc.py [hunk 3]
```diff
@@ -61,3 +61,27 @@ def get_document_url(self, request):
...
+        """Get the configuration of an assignment that was original deep linked."""
+        params = {}
+        possible_parameters = ["url", "group_set"]
+
+        for param in possible_parameters:
+            # Get the value from the custom parameters set during deep linking
+            if value := request.lti_params.get(f"custom_{param}"):
+                params[param] = value
+
+        return params
```
Comment (jon-betts) [inline lms/product/plugin/misc.py:87]: I feel like this is premature. We only have two examples and they are totally different?

I think the assumption here is that D2L is doing it right.

File: lms/views/lti/deep_linking.py (hunk 2 above)
Comment (jon-betts) [inline lms/views/lti/deep_linking.py:131]:
```suggestion
                    # The URL we will be called back on when the
                    # assignment is launched from the LMS
                    "url": self.misc_plugin.get_deeplinking_launch_url(
                        self.request, assignment_configuration
                    ),
                    # These values should be passed back to us as custom
                    # LTI params, but Canvas doesn't seem to.
                    "custom": assignment_configuration,
```

File: lms/views/lti/deep_linking.py [hunk 4]
```diff
@@ -179,22 +185,19 @@ def file_picker_to_form_fields_v11(self):
                         {
                             "@type": "LtiLinkItem",
                             "mediaType": "application/vnd.ims.lti.v1.ltilink",
-                            "url": url,
+                            "url": self.misc_plugin.get_deeplinking_launch_url(
+                                self.request, assignment_configuration
+                            ),
+                            "custom": assignment_configuration,
```
Comment (jon-betts) [inline lms/views/lti/deep_linking.py:191]:
```suggestion
                            # The URL we will be called back on when the
                            # assignment is launched from the LMS
                            "url": self.misc_plugin.get_deeplinking_launch_url(
                                self.request, assignment_configuration
                            ),
                            # These values should be passed back to us as custom
                            # LTI params, but Canvas doesn't seem to.
                            "custom": assignment_configuration,
```

File: lms/views/lti/deep_linking.py [hunk 5]
```diff
@@ -145,7 +151,7 @@ def file_picker_to_form_fields_v13(self):
             LTIEvent(
                 request=self.request,
                 type=LTIEvent.Type.DEEP_LINKING,
-                data={"document_url": document_url},
+                data=assignment_configuration,
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/views/lti/deep_linking.py:154]: Do we want a migration to normalize the old ones in any way?

**Review by jon-betts — APPROVED** [review]
This works and the testing instructions are good.

I think the plugins here are going in the wrong direction, but we might have to go through some mess to get out the other side to something good. As a result I'm not really going to provide many suggestions about the factoring here.

We are getting a proliferation of methods which are arranged as `get_<particular_assignment_value>_in_<particular_situation()`. I think we probably want to take a swing at a `get_settings()` methods in the assignment service which hides all the details.

This could:

 * Hide all the details of deeplinking vs. not
 * Hide all the details of course copy
 * Potentially suck a bunch of the upserting logic out of the launch views

My gut is telling me to return a settings data-class, not the raw assignment object. I think talking in this medium would simplify the interface and might make the plugins easier to write.

**Pull request merged**